src/test/java/com/nique1/musicmanagementsystem/song/application/api/SongControllerTest.java

```diff
@@ -177,14 +181,206 @@ void shouldReturnSongDataGivenTrackAndYearExists() throws Exception {
+                .andExpect(jsonPath("$[0].trackLength").value(295))
+                .andExpect(jsonPath("$[0].year").value(2010))
+                .andExpect(jsonPath("$[1].uuid").value("82f81b56-aa8f-4850-8bca-f75a12e6ea57"))
+                .andExpect(jsonPath("$[1].artistName").value("Nirvana"))
+                .andExpect(jsonPath("$[1].trackName").value("Smells Like Teen Spirit"))
+                .andExpect(jsonPath("$[1].trackLength").value(301))
+                .andExpect(jsonPath("$[1].year").value(1990))
+                .andExpect(jsonPath("$[2].uuid").value("8b477224-d930-402b-b284-cfd1b59bba86"))
+                .andExpect(jsonPath("$[2].artistName").value("Michael Jackson"))
+                .andExpect(jsonPath("$[2].trackName").value("Billie Jean"))
+                .andExpect(jsonPath("$[2].trackLength").value(294))
+                .andExpect(jsonPath("$[2].year").value(1980))
+                .andExpect(jsonPath("$[3].uuid").value("58745f5f-937f-412b-befd-cb55acb8e70e"))
+                .andExpect(jsonPath("$[3].artistName").value("Led Zeppelin"))
+                .andExpect(jsonPath("$[3].trackName").value("Stairway to Heaven"))
+                .andExpect(jsonPath("$[3].trackLength").value(482))
+                .andExpect(jsonPath("$[3].year").value(1970))
+                .andExpect(jsonPath("$[4].uuid").value("f0955f27-4404-49c7-b444-a5c3f1918b52"))
+                .andExpect(jsonPath("$[4].artistName").value("The Beatles"))
+                .andExpect(jsonPath("$[4].trackName").value("Come Together"))
+                .andExpect(jsonPath("$[4].trackLength").value(259))
+                .andExpect(jsonPath("$[4].year").value(1960));
+
+    }
+
+
+    @Test
+    void shouldReturnEmptyListGivenNullParameters() throws Exception {
+        //given
+        given(songFacade.getSongByArtistTrackOrYear(null, null, null, null)).willReturn(List.of());
+        //when-then
+        mockMvc.perform(get("/songs"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$").isEmpty());
+    }
+
+    @Test
+    void shouldReturnEmptyListGivenArtistNameNotFound() throws Exception {
+        //given
+        given(songFacade.getSongByArtistTrackOrYear("Some artist", null, null, null)).willReturn(List.of());
+        //when-then
+        mockMvc.perform(get("/songs"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$").isEmpty());
+    }
+
+    @Test
+    void shouldReturnEmptyListGivenTrackNameNotFound() throws Exception {
+        //given
+        given(songFacade.getSongByArtistTrackOrYear(null, "Some track", null, null)).willReturn(List.of());
+        //when-then
+        mockMvc.perform(get("/songs"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$").isEmpty());
+    }
+
+    @Test
+    void shouldReturnSongDataGivenPartialLowercaseArtistName() throws Exception {
```

Comment:
Ten przypadek i te dalsze mają sens głównie na poziomie testów repozytorium, bo tutaj i tak tylko przekazujesz dane do fasady, a wyszukiwanie i obsługa tego dzieje się głębiej.

To nie znaczy, że takie przypadki testowe zupełnie nie mają zastosowania, ale test trzeba byłoby przygotować w inny sposób. Jeśli masz chwilę, możesz w ramach ćwiczenia przygotować równoległą klasę testową `SongControllerIntegrationTest` gdzie już nie mockujesz fasady, za to masz posadzony pełen kontekst springowy i bazę danych w kontenerze (tak samo jak w `SongRepositoryAdapterTest`) i tam te testy, w dokładnie takiej formie (z pominięciem sekcji _given_) wprowadzić.

Wtedy będzie to taki pełen test integracyjny, gdzie testujemy całą ścieżkę od kontrolera, do bazy danych.